python/ray/serve/deployment_executor_node.py

```diff
@@ -0,0 +1,78 @@
+from typing import Any, Dict, List
+
+from ray.experimental.dag import DAGNode
+from ray.serve.deployment_method_executor_node import DeploymentMethodExecutorNode
+from ray.experimental.dag.constants import DAGNODE_TYPE_KEY, PARENT_CLASS_NODE_KEY
+from ray.experimental.dag.format_utils import get_dag_node_str
+
+
+class DeploymentExecutorNode(DAGNode):
+    """The lightweight executor DAGNode of DeploymentNode that optimizes for
+    efficiency.
+
+        - We need Ray DAGNode's traversal and replacement mechanism to deal
+            with deeply nested nodes as args in the DAG
+        - Meanwhile, __init__, _copy_impl and _execute_impl are on the critical
+            pass of execution for every request.
+
+    Therefore for serve we introduce a minimal weight node as the final product
+    of DAG transformation, and will be used in actual execution as well as
+    deployment.
+    """
+
+    def __init__(
+        self,
+        deployment_handle,
+        dag_args,  # Not deployment init args
+        dag_kwargs,  # Not deployment init kwargs
+    ):
+        self._deployment_handle = deployment_handle
+        super().__init__(dag_args, dag_kwargs, {}, {})
+
+    def _copy_impl(
+        self,
+        new_args: List[Any],
+        new_kwargs: Dict[str, Any],
+        new_options: Dict[str, Any],
+        new_other_args_to_resolve: Dict[str, Any],
+    ):
+        return DeploymentExecutorNode(
+            self._deployment_handle,
+            new_args,
+            new_kwargs,
+        )
+
+    def _execute_impl(self, *args, **kwargs):
+        """Does not call into anything or produce a new value, as the time
+        this function gets called, all child nodes are already resolved to
+        ObjectRefs.
+        """
+        return self._deployment_handle
```

Comment:
I'm confused at what `_execute_impl` does... should this just be calling the deployment handle? I thought the contract of `_execute_impl` was that it always returns an `ObjectRef`?

Reply:
Actually, is it even valid for this to be executed directly? I think we should only have deployment method nodes at time of execution, right?

Reply:
We need to have a node that captures class / deployment creation. For Ray Core DAG, this is creating a new actor; For Serve's `DeploymentNode`, we deferred the work to subsequent `Deployment` objects to be deployed in `serve.run()`. 

So in the minimal executor dag version of `DeploymentNode`, it should do literally nothing but returning the handle, but it's also necessary because we need to support user passing `DeploymentNode` around into other func or classes' constructor. 

So a simpler way to understand this function is something like 

```
@serve.deployment
class Combine:
    def __init__(self, model):
        self.model = model
``` 

Where the `model` here is a `DeploymentNode`, and we swapped it with a `DeploymentExecutorNode` that upon execution returns the handle to `model` for init args.

Reply:
Why is this "upon execution?" Shouldn't this just be in the initial transformation step where we go from `DeploymentNode` -> `Handle`? Not understanding why the intermediate transformation to `DeploymentExecutorNode` is necessary.

Reply:
My mental model of how DAG execution work is just to start from the output node, and recursively call the execute_impl of its dependencies, and replace the dependencies with the result of the call. 

for example, MethodNode's first calls its parent DeployeNode exec_impl, which returns the handle, and then the MethodNode's `parent_node` _becomes_ the handle object, so the method node's impl can be something like `self.parent_node.method_name.remote()`.